### PR TITLE
Wrap markdown rendering in iframe

### LINKED ISSUE #762: Potential security problems through custom css in document.
**Description**
Right now the markdown parser doesn't delete custom css. This could to security problems, because you can manipulate the style of the whole page.

**Expected behavior**
One of these:
- Deactivate custom CSS (not prefered, because it can be useful)
- Apply CSS only in renderer (would be nice)
- Turn custom css on / off (not prefered, because you just turn the bug on and off)

**Screenshots**
image

### PR DESCRIPTION
**Pull request by mrdrogdrog**
### Component/Part
Markdown renderer

### Description
This PR wrappes the render output into an iframe to avoid css injections.

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] Extended changelog
- [x] I read the [contribution documentation](https://github.com/hedgedoc/react-client/blob/main/CONTRIBUTING.md) and signed-off my commits to accept the DCO.

### Related Issue(s)
Fixes #762

### CODE REVIEW

File: src/components/editor/document-renderer-pane/document-iframe.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import React, { useCallback, useRef, useState } from 'react'
```
Comment (DerMolly) [inline src/components/editor/document-renderer-pane/document-iframe.tsx:1]:
```suggestion
/*
SPDX-FileCopyrightText: 2020 The HedgeDoc developers (see AUTHORS file)
SPDX-License-Identifier: AGPL-3.0-only
*/

import React, { useCallback, useRef, useState } from 'react'
```

**Review by DerMolly — CHANGES_REQUESTED** [review]

**Comment by DerMolly** [comment]
Not working with this:

- [ ] scroll-sync
- [ ] vega-lite dots are huge
- [ ] mermaid
- [ ] the toast of the copy-button in code blocks (it appears somewhere over the editor)

**Comment by mrdrogdrog** [comment]
- [ ] Prevent navigation in iframe
- [ ] Open links in new tabs
- [ ] Don't open tabs for jump marks

**Comment by DerMolly** [comment]
image
These buttons don't work.

**Comment by mrdrogdrog** [comment]
> image
> These buttons don't work.

Fixed

**Review by DerMolly — APPROVED** [review]

**Review by DerMolly — APPROVED** [review]

**Review by DerMolly — APPROVED** [review]

File: src/components/markdown-renderer/replace-components/image/image-replacer.tsx [hunk 2]
```diff
@@ -9,17 +9,27 @@ import React from 'react'
...
+    super()
+    this.clickHandler = clickHandler
+  }
+
   public getReplacement (node: DomElement): React.ReactElement | undefined {
     if (node.name === 'img' && node.attribs) {
       return <ProxyImageFrame
         id={node.attribs.id}
-        className={node.attribs.class}
+        className={`${node.attribs.class} cursor-pointer`}
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/replace-components/image/image-replacer.tsx:26]: Didn't we want to use the zoom-in cursor here?

**Review by ErikMichelson — CHANGES_REQUESTED** [review]
I don't know whether it's intended, but the wide embeddings in view-mode don't seem to work anymore (as they're not wide).

File: src/components/render-page/iframe-communicator.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/*
+ * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+export abstract class IframeCommunicator<SEND, RECVIEVE> {
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply src/components/render-page/iframe-communicator.ts]: fixed

File: src/components/markdown-renderer/replace-components/image/image-replacer.tsx (hunk 2 above)
Comment (mrdrogdrog) [reply src/components/markdown-renderer/replace-components/image/image-replacer.tsx]: fixed

File: src/components/editor/splitter/splitter.tsx [hunk 4]
```diff
@@ -33,28 +33,39 @@ export const Splitter: React.FC<SplitterProps> = ({ containerClassName, left, ri
     setSplit(newSize * 100)
   }
 
+  const onMouseUp = useCallback(() => {
+    setDoResizing(false)
+  }, [])
+
+  const onTouchEnd = useCallback(() => {
+    setDoResizing(false)
+  }, [])
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply src/components/editor/splitter/splitter.tsx:42]: fixed

File: cypress/support/getMarkdownRenderer.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+declare namespace Cypress {
+  interface Chainable {
+    /**
+     * Custom command to check an external Link.
+     * @example cy.get(a#extern).checkExternalLink('http://example.com')
+     */
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply cypress/support/getMarkdownRenderer.ts]: I removed the docs

**Review by ErikMichelson — APPROVED** [review]

**Review by DerMolly — APPROVED** [review]

**Pull request merged**